File: MAY 24/21 May/K closest elements gfg.cpp
class Solution {
  public:
    vector<int> printKClosest(vector<int> arr, int n, int k, int x) {
        // code here
        vector<int> closest;
        int lb = lower_bound(arr.begin(), arr.end(), x) - arr.begin();
        int left = lb-1, right = lb;
        if(lb<n && arr[lb]==x)
            right = lb+1;
        
        int lDiff, rDiff;
        while(k && left>=0 && right<n)
        {
            lDiff = x - arr[left];
            rDiff = arr[right] - x;
            if(lDiff<rDiff)
            {
                closest.push_back(arr[left]);
                left--;
            }
            else
            {
                closest.push_back(arr[right]);
                right++;
            }
            k--;
        }
        while(k && left>=0)
        {
            closest.push_back(arr[left]);
            left--;
            k--;
        }
        while(k && right<n)
        {
            closest.push_back(arr[right]);
            right++;
            k--;
        }
        return closest;
    }
};

File: JUNE 2025/12 June/K closest elements gfg.java
class Solution {
    int lowerBound(int[] arr, int x)
    {
        int low=0, high=arr.length-1, mid;
        while(low<high)
        {
            mid = low+(high-low)/2;
            if(arr[mid]<x)
                low = mid+1;
            else
                high = mid;
        }
        return low;
    }
    int[] printKClosest(int[] arr, int k, int x) {
        // code here
        int index = lowerBound(arr, x);
        int N = arr.length;
        int left=index-1, right=index;
        if(index<N && arr[index]==x)
            right++;
        
        ArrayList<Integer> closest = new ArrayList<>();
        while(left>=0 && right<N && k!=0)
        {
            if(Math.abs(x-arr[left])>=Math.abs(x-arr[right]))
            {
                closest.add(arr[right++]);
            }
            else
                closest.add(arr[left--]);
            k--;
        }
        while(left>=0 && k!=0)
        {
            closest.add(arr[left--]);
            k--;
        }
        
        while(right<N && k!=0)
        {
            closest.add(arr[right++]);
            k--;
        }
        
        return closest.stream().mapToInt(Integer::intValue).toArray();
    }
}
